src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -769,90 +670,268 @@ private static FunctionCompletedMessage CreateCompletedMessage(FunctionStartedMe
                 ParameterLogBlob = startedMessage.ParameterLogBlob
             };
         }
-
-        private static IDictionary<string, string> CreateArguments(IReadOnlyDictionary<string, IValueProvider> parameters)
+                
+        // Handle various phases of parameter building and logging.         
+        internal class ParameterHelper
         {
-            IDictionary<string, string> arguments = new Dictionary<string, string>();
+            // Logs, contain the result from invoking the IWatchers.
+            private IDictionary<string, ParameterLog> _parameterLogCollector = new Dictionary<string, ParameterLog>();
+
+            // Optional runtime watchers for the parameters. 
+            private IReadOnlyDictionary<string, IWatcher> _parameterWatchers;
 
-            if (parameters != null)
+            // ValueProviders for the parameters. These are produced from binding. 
+            private IReadOnlyDictionary<string, IValueProvider> _parameters;
+
+            // Phsyical objects to pass to the underlying method Info. These will get updated for out-parameters. 
+            // These are produced by executing the binders. 
+            private object[] _invokeParameters;
+
+            // ordered parameter names of the underlying physical MethodInfo that will be invoked. 
+            private IReadOnlyList<string> _parameterNames;             
+
+            public ParameterHelper(IReadOnlyList<string> parameterNames)
             {
-                foreach (KeyValuePair<string, IValueProvider> parameter in parameters)
-                {
-                    arguments.Add(parameter.Key, parameter.Value.ToInvokeString());
-                }
+                this._parameterNames = parameterNames;
             }
 
-            return arguments;
-        }
+            public IDictionary<string, ParameterLog> ParameterLogCollector => _parameterLogCollector;
 
-        private static int GetParameterIndex(IReadOnlyList<string> parameterNames, string name)
-        {
-            for (int index = 0; index < parameterNames.Count; index++)
+            public object[] InvokeParameters => _invokeParameters;
+
+            internal static ParameterHelper New(object[] args)
```

Comment:
It's confusing that this construction mechanism takes an array of arg values, while your other constructor just takes param names. Which should be used when?